File: P1/Ejercicio1/factoria/FactoriaCarretera.java
package factoria;
import bicicleta.Bicicleta;
import bicicleta.BicicletaCarretera;
import carrera.Carrera;
import carrera.CarreraCarretera;

/**
 * Clase que representa la factoría que se encargará de crear las carreras y las bicicletas de carretera. Esta
 * implementa la interfaz FactoriaCarreraYBicicleta
 */
public class FactoriaCarretera implements FactoriaCarreraYBicicleta {
    /**
     * Crea una carrera de bicicletas de carretera
     * @param numBicicletas Número de bicicletas de carretera que contendrá la carrera
     * @return Objeto Carrera con la carrera de carretera
     */
    @Override
    public Carrera crearCarrera(int numBicicletas) {
        return new CarreraCarretera(numBicicletas);
    }

    /**
     * Crea una bicicleta de carretera
     * @param id Identificador que tendrá la bicicleta de carretera
     * @return Objeto Bicicleta con la bicicleta de carretera
     */
    @Override
    public Bicicleta crearBicicleta(int id) {
        return new BicicletaCarretera(id);
    }
}

File: P1/Ejercicio2/Factorias/factoria_Carretera.py
from Carreras.carrera_Carretera import CarreraCarretera
from Bicicletas.bicicleta_Carretera import BicicletaCarretera
import random
from Factorias.factoriaCarreraYBicicleta import FactoriaCarreraYBicicleta

class FactoriaCarretera(FactoriaCarreraYBicicleta):
    """
    Implementa la fábrica concreta para crear carreras y bicicletas de carretera.
    Sigue el patrón de fábrica abstracta, método abstracto y prototipo.
    
    Atributos:
        prototipo_bicicleta (BicicletaCarretera): Un prototipo de bicicleta de carretera para clonar nuevas bicicletas.
        
    Métodos:
        crear_item: Se crean carreras de carretera y bicicletas de carretera, estas última por clonación.
    """
    
    def __init__(self):
        """
        Inicialización del prototipo de bicicleta para clonación.
        """
        super().__init__(BicicletaCarretera(0, 0))   # Prototipo específico para FactoriaCarretera  

    def crear_item (self, item, id):
        """
        Se crean carreras de carretera y bicicletas de carretera, estas última por clonación y ajustando su id y puntuacion.
                
        Parametros:
            item: cadea de texto que describe el objeto a crear, una carrera o una bicicleta.
        """
        if item == 'carrera':
            return CarreraCarretera()
        elif item == 'bicicleta':
            bicicleta_clonada = self._prototipo_bicicleta.clone()
            bicicleta_clonada._id = id  # Usa el nuevo id
            bicicleta_clonada._puntuacion = random.randint(1, 20)
            return bicicleta_clonada